src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,497 @@
+    """
+    A class to represent all Container Registries other than Azure Container Registry
+    and handle all Registry operations (primarily using docker CLI).
+    """
+
+    def find_image(
+        self, image: str, version: str
+    ) -> Union[Tuple[ContainerRegistry, str], Tuple[None, None]]:
+        """
+        Find whether the given image exists in this registry.
+
+        :param image: The image to find in the registry
+            (the image should be stripped of all namespace)
+        :param version: The version of the image (also known as image tag)
+        :return: The registry and namespace for the image
+        """
+
+        for namespace in self.registry_namespaces:
+            image_path = f"{self.registry_name}/{namespace}{image}:{version}"
+
+            logger.debug(
+                "Checking if image %s:%s exists in %s",
+                image,
+                version,
+                self.registry_name,
+            )
+
+            try:
+                manifest_inspect_cmd = [
+                    str(shutil.which("docker")),
+                    "manifest",
+                    "inspect",
+                    image_path,
+                ]
+
+                output = call_subprocess_raise_output(manifest_inspect_cmd)
+
+                if output is not None:
+                    return self, namespace
+
+            except CLIError as error:
+                if "manifest unknown" in str(error):
+                    continue
+                logger.warning(
+                    (
+                        "Failed to contact source registry %s. "
+                        "Make sure you run docker login on this registry "
+                        "before running the aosm command."
+                    ),
+                    self.registry_name,
+                )
+                logger.debug(error, exc_info=True)
+                raise ClientRequestError(
+                    f"Failed to contact source registry {self.registry_name} "
+                    "Make sure you run docker login on this registry "
+                    "before running the aosm command."
+                ) from error
+        return None, None
+
+
+class AzureContainerRegistry(ContainerRegistry):
+    """
+    A class to represent an Azure Container Registry and handle all ACR operations.
+    """
+
+    # TODO: Have I tested this code path?
+    def _login(self) -> None:
+        """
+        Log in to the source registry using the Azure CLI.
+        Uses the CLI user's context to log in to the source registry.
+        """
+
+        message = f"Logging into source registry {self.registry_name}"
+        logger.info(message)
+
+        try:
+            acr_source_login_cmd = [
+                str(shutil.which("az")),
+                "acr",
+                "login",
+                "--name",
+                self.registry_name,
+            ]
+            call_subprocess_raise_output(acr_source_login_cmd)
+
```

Comment:
I think generally we don't leave blank lines between `try` and `except` blocks. Have you run `black` over this?

Reply:
I did, it did not change it but I can change it